File: Java/0155_Min_Stack.java
import java.util.Stack;

// 2020-08-07 17:25:37

class MinStack {


    private Stack<Integer> stack = new Stack<>();
    private Stack<Integer> min = new Stack<>();

    /** initialize your data structure here. */
    public MinStack() {}
    
    public void push(int x) {
        stack.push(x);
        if (min.isEmpty() || x <= getMin()) min.push(x);
    }
    
    public void pop() {
        if (stack.peek() == getMin()) min.pop();
        stack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Python/0155_min_stack.py
# 2020-08-07 23:52:39

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min_ = []
        self.stack = []
        

    def push(self, x: int) -> None:
        self.stack.append(x) 
        if len(self.min_) ==  0 or self.getMin() >= x:
            self.min_.append(x)


    def pop(self) -> None:
        if self.min_[-1] == self.top():
            self.min_.pop()
        self.stack.pop()


    def top(self) -> int:
        return self.stack[-1]
        

    def getMin(self) -> int:
        return self.min_[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()